src/middlewared/middlewared/plugins/sysdataset.py

```diff
@@ -500,33 +502,57 @@ async def setup_datasets(self, pool, uuid):
                 except Exception:
                     self.logger.warning("Failed to replace dataset [%s].", dataset, exc_info=True)
             else:
-                update_props_dict = {k: {'value': v} for k, v in props.items()
-                                     if datasets_prop[dataset][k]['value'] != v}
+                update_props_dict = {
+                    k: {'value': v} for k, v in props.items()
+                    if datasets_prop[dataset][k]['value'] != v
+                }
                 if update_props_dict:
                     await self.middleware.call(
                         'zfs.dataset.update',
                         dataset,
                         {'properties': update_props_dict},
                     )
 
+            try:
+                await self.middleware.run_in_thread(self.__create_relevant_paths, config.get('create_paths', []))
+            except Exception:
+                self.logger.error('Failed to create relevant paths for %r', dataset, exc_info=True)
+
+    def __create_relevant_paths(self, create_paths):
+        for create_path_config in create_paths:
+            os.makedirs(create_path_config['path'], exist_ok=True)
+            cpath_stat = os.stat(create_path_config['path'])
+            if all(create_path_config[k] for k in ('uid', 'gid')) and (
+                cpath_stat.st_uid != create_path_config['uid'] or cpath_stat.st_gid != create_path_config['gid']
+            ):
+                os.chown(create_path_config['path'], create_path_config['uid'], create_path_config['gid'])
+
     def __mount(self, pool, uuid, path=SYSDATASET_PATH):
         """
         Mount group of datasets associated with our system dataset.
         `path` will be either  SYSDATASET_PATH or temp dir in the middlewared
         rundir. The latter occurs when migrating dataset between pools.
         """
         mounted = False
-        for dataset, name in self.__get_datasets(pool, uuid):
-            if name:
-                mountpoint = f'{path}/{name}'
+        for ds_config in get_system_dataset_spec(pool, uuid):
+            dataset, name = ds_config['name'], os.path.basename(ds_config['name'])
+            if desired_mountpoint := ds_config.get('mountpoint'):
+                mountpoint = desired_mountpoint
             else:
-                mountpoint = path
+                mountpoint = f'{path}/{name}'
+
             if os.path.ismount(mountpoint):
                 continue
 
             with suppress(FileExistsError):
                 os.mkdir(mountpoint)
             subprocess.run(['mount', '-t', 'zfs', dataset, mountpoint], check=True)
+
+            if chown_config := ds_config.get('chown_config'):
```

Comment:
Let's default to `-1` for uid and gid in the chown config. Then we can skip the stat check and just chown. -1 here has special meaning to leave existing value.

Reply:
@anodos325 to be sure i follow - for all datasets we have a default `chown_config` that where ever they get mounted we do `chown` but using `-1`?

Reply:
In terms of operations - it would mean that we always do `chown` so another extra step..not sure if it has performance impacts as we are using `-1`

Reply:
@Qubad786 -1 means the OS won't change the ownership values if they're the same so it actually reduces I/O.